rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,110 @@
+
+In order to know which `$out`s refer to a particular `$cas`, symlinks can be used (`$out` pointing to `$cas`), or that data can be stored in the store database. The database can help with doing reverse lookups from `$cas` to all the `$out`s. Using symlinks will have a benefit of handling the case when self-references that are not discoverable via grep (e.g. filtered by ```xxd(1)```).
+
+## Calculating `$cas`
+
+There is one important corner case that needs special handling: if a derivation refers to itself, it will be referring to `$out`, because `$cas` is not known at the time of the build. This means that each `$out` of a furthermore equivalent build would have a different hash, due to the different `$out`s.
+
+To fix this, the `$cas` calculation has to replace all occurrences of `$out` with an equal-length string of (for example) NULL bytes. After that, `$out` is renamed to `$cas` and all occurrences of `$out` are replaced with `$cas`.
+
+This also means that `$out` and `$cas` should have the same length. The easiest way to achieve that is to use the same hash function for the output hash as used for the input hashes.
+
+To calculate `$cas` we need to include all the data that uniquely defines a derivation: the file contents, case-sensitive names, and the permission bits, traversed in a fixed order, no matter what the filesystem or platform. Not to be included are the owning `uid` of the store and timestamps.
+
+## Distributing derivations
+
+Since `$cas` is only known when `$out` is built, binary caches would need to retain that information. When you look up `$out` to see if it was built already, the response should be _"Yes, this is available as `$cas`"_.
+
+## Maintaining the Nix store
+
+When garbage collecting, the Nix store should also remove `$out` references (be they symlinks or db entries) when removing a `$cas`.
+
+## Micro-optimizations not worth considering
+
+* By stripping the version from `$cas`, it could be the same for multiple versions of the same derivation.
+  * However, increased version numbers mean the derivation actually changed, so there is no point in doing that.
+* By stripping the name and the version from `$cas`, it could be the same for multiple different derivations.
+  * However, this makes it hard to find out what derivation a certain `$cas` is
+  * Furthermore, different inputs with the same contents are very unlikely, and there is no reduction in builds that need to be done.
+
+Finally, `nix-store` supports hardlinking duplicate files, so the above optimizations are useless.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Extra code to maintain
+* Slightly more processing after a build
```

Comment:
Most likely, the implementation will restrict `$out`-referring compressed archives (JARs etc.); right now they don't really create problems (and the dependency references there are OK as long as the dependencies are also mentioned outside the compressed part at least once)

Reply:
Hmm indeed, that is problematic. Do you know of any jars that refer to their own $out?
The only way this can work is by calculating the CAS before compression :-/

Reply:
the problem is indeed observable — I checked compressed man pages, because I have more of them.

`/nix/store/5fx36jkalfnkwfq1w4hn2nqmixzi0r8d-inkscape-0.92.3/share/man/man1/inkview.1.gz:\&    inkview B</nix/store/5fx36jkalfnkwfq1w4hn2nqmixzi0r8d\-inkscape\-0.92.3/share>/inkscape/examples/*.svg*
`

`/nix/store/ns8fcnqz4dpjc5z24nmxnrgr0qbj8321-vim_configurable-8.1.0578/share/man/ru.UTF-8/man1/vim.1.gz:/nix/store/ns8fcnqz4dpjc5z24nmxnrgr0qbj8321-vim_configurable-8.1.0578/share/vim/vim81/bugreport.vim`

etc.

Technically speaking, man page compression is something our stdenv knows about, so adding some secial support for manpages is probably an option…

Reply:
For the record (and in case anyone finds it during a more general dependency-rewriting discussion): LibreOffice has jars with dependency references (Bean shell) but not self-references.

Reply:
Now that you mention it, maybe Mono is worth checking for problems related to that…

Reply:
FWIW, I think part of the answer here is to have a primitive way to tell Nix what the dependencies of a build are, and move hash-scanning into stdenv as a default path but allowing custom scanning or reporting per derivation.

Reply:
Re: finding dependencies in build-time, a package can already do as you say and put the list into `nix-support`? Not sure what should be added from dependency-finding perspective. In any case dependencies are easier because you need to find at least one references, not all of them…

Intensional store via path rewriting might indeed require pluggable extra rewriters (specified by derivation).

Reply:
@7c6f434c Sure, but you can't turn off scanning (or use a better version) or remove references this way, and putting it into the output is kind of silly. The idea (which is really separate from this RFC) is to separate mechanism and policy.

As for pluggable rewriters, I'm a huge fan of that idea as well. Would love to move *everything* related to hash scanning out of nix core and into user nix code.

Reply:
I think that _adding_ better scanning passes and not being able to ignore references is acceptable per se.

Of course I am in favour of any flexible design with wide possibilities of overrides, but different overrides provide different amount of value.

(It's separately annoying how all the dependency substitution work would become much simpler if bind mounts for non-root users were truly universally available (directory hardlinks would be fine, too) — then the dependency replacement would be about just changing a few symlinks inside the package once, as the package's build system hopefully wouldn't be able to resolve the bind mounts)

Reply:
By using only `$cas` dependencies, the only paths that need rewriting are self-references. Maybe we need pluggable rewriters, but it is my hope that we can make all builds work as-is with only string replacements.